internal/states/encryption/flow/interface.go

```diff
@@ -0,0 +1,102 @@
+// Package flow contains the top-level flow for client-side state encryption and the interfaces an encryption
+// or key derivation method need to implement.
+package flow
+
+import "github.com/opentofu/opentofu/internal/states/encryption/encryptionconfig"
+
+type ConfigurationSource string
+
+const (
+	ConfigurationSourceEnvDefault ConfigurationSource = "env-default"
+	ConfigurationSourceCode       ConfigurationSource = "code"
+	ConfigurationSourceEnv        ConfigurationSource = "env"
+)
+
+// Flow represents the top-level state or plan encryption/decryption flow
+// for a particular encryption configuration.
+//
+// Instances of Flow are kept in an internal cache per configuration key.
+// Unless you are writing tests, you should not create them in your code.
+//
+// See also encryption.Instance().
+type Flow interface {
+	// DecryptState decrypts encrypted state.
+	//
+	// If the state is not actually encrypted, it will be returned
+	// as-is. DecryptState will first try decryption using the state
+	// encryption configuration. If this fails, it tries the decryption
+	// fallback configuration. If neither produces a valid result it fails.
+	//
+	// payload must be a json document passed in as a []byte.
+	// This can be encrypted or unencrypted state.
+	//
+	// If no error is returned, then the first return value will always
+	// be a json document, possibly the same one passed in as payload
+	// if the payload was not actually encrypted.
+	DecryptState(payload []byte) ([]byte, error)
```

Comment:
Is there a reason we specifically wanted DecryptState AND DecryptPlan instead of just a generic Encrypt/Decrypt methods?

Reply:
I tried to lump these two together like I did in the PoC, but then I added the comments and they became really unclear.

I think this is probably an indication these should be separate functions on the outward facing interface.

The differences are (maybe need to be clearer about those in the comments):
- plaintext planfile is binary (zip format) while state is json
- planfile does not need a decryption fallback (not retained long term) while state does
- planfile can only use full encryption while state can use all encryption methods

Other than easier-to-read comments and differences in validation, the two implementations will share most of their functionality.